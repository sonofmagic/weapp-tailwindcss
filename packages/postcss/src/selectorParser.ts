export { isOnlyBeforeAndAfterPseudoElement } from './selectorParser/before-after'

export { getFallbackRemove } from './selectorParser/fallback'

export { ruleTransformSync } from './selectorParser/rule-transformer'

export {
  composeIsPseudoAst,
  getCombinatorSelectorAst,
  mklist,
} from './selectorParser/utils'
