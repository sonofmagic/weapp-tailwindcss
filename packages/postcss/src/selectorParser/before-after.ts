import type { Rule } from 'postcss'
import psp from 'postcss-selector-parser'

interface BeforeAfterState {
  before: boolean
  after: boolean
}

let beforeAfterStateRef: BeforeAfterState | null = null

const beforeAfterParser = psp((selectors) => {
  const state = beforeAfterStateRef
  if (!state) {
    return
  }
  selectors.walkPseudos((s) => {
    if (s.parent?.length === 1) {
      if (/^:?:before$/.test(s.value)) {
        state.before = true
      }
      if (/^:?:after$/.test(s.value)) {
        state.after = true
      }
    }
  })
})

export function isOnlyBeforeAndAfterPseudoElement(node: Rule) {
  const state: BeforeAfterState = {
    before: false,
    after: false,
  }
  beforeAfterStateRef = state
  beforeAfterParser.astSync(node)
  beforeAfterStateRef = null
  return state.before && state.after
}
