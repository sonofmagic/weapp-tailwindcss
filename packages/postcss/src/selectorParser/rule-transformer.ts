import type { Rule } from 'postcss'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'
import { composeIsPseudo, internalCssSelectorReplacer } from '../shared'
import {
  getCombinatorSelectorAst,
  normalizeTransformOptions,
} from './utils'

export type RuleTransformer = (rule: Rule) => void

const ruleTransformCache = new WeakMap<IStyleHandlerOptions, RuleTransformer>()

function createRuleTransformer(options: IStyleHandlerOptions): RuleTransformer {
  let currentRule: Rule | undefined

  const { escapeMap, cssSelectorReplacement, cssRemoveHoverPseudoClass, uniAppX } = options

  const transform: SyncProcessor = (selectors) => {
    const rule = currentRule
    if (!rule) {
      return
    }

    selectors.walk((selector, index) => {
      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
        })
      }
      else if (selector.type === 'universal') {
        if (cssSelectorReplacement?.universal) {
          selector.value = composeIsPseudo(cssSelectorReplacement.universal)
        }
      }
      else if (selector.type === 'selector') {
        if (cssRemoveHoverPseudoClass) {
          const node = selector.nodes.find(x => x.type === 'pseudo' && x.value === ':hover')
          if (node) {
            selector.remove()
          }
        }
      }
      else if (selector.type === 'pseudo') {
        if (
          selector.value === ':root'
          && cssSelectorReplacement?.root) {
          selector.value = composeIsPseudo(cssSelectorReplacement.root)
        }
        else if (selector.value === ':where') {
          if (uniAppX) {
            selector.value = ':is'
          }
          if (index === 0 && selector.length === 1) {
            selector.walk((node, idx) => {
              if (idx === 0 && node.type === 'class') {
                const nodes = node.parent?.nodes
                if (nodes) {
                  const first = nodes[idx + 1]
                  if (first && first.type === 'combinator' && first.value === '>') {
                    const second = nodes[idx + 2]
                    if (second && second.type === 'pseudo' && second.value === ':not' && second.first.first.type === 'pseudo' && second.first.first.value === ':last-child') {
                      const ast = getCombinatorSelectorAst(options)

                      second.replaceWith(
                        ...ast,
                      )
                    }
                  }
                }
              }
            })

            selector.replaceWith(...selector.nodes)
            for (const node of rule.nodes) {
              if (node.type === 'decl') {
                if (node.prop === 'margin-block-start') {
                  node.prop = 'margin-block-end'
                }
                else if (node.prop === 'margin-block-end') {
                  node.prop = 'margin-block-start'
                }
                else if (node.prop === 'margin-inline-start') {
                  node.prop = 'margin-inline-end'
                }
                else if (node.prop === 'margin-inline-end') {
                  node.prop = 'margin-inline-start'
                }
                else if (node.prop === 'margin-top') {
                  node.prop = 'margin-bottom'
                }
                else if (node.prop === 'margin-bottom') {
                  node.prop = 'margin-top'
                }
                else if (node.prop === 'margin-left') {
                  node.prop = 'margin-right'
                }
                else if (node.prop === 'margin-right') {
                  node.prop = 'margin-left'
                }
                else if (node.prop === 'border-inline-start-width') {
                  node.prop = 'border-right-width'
                }
                else if (node.prop === 'border-inline-end-width') {
                  node.prop = 'border-left-width'
                }
                else if (node.prop === 'border-top-width') {
                  node.prop = 'border-bottom-width'
                }
                else if (node.prop === 'border-bottom-width') {
                  node.prop = 'border-top-width'
                }
                else if (node.prop === 'border-left-width') {
                  node.prop = 'border-right-width'
                }
                else if (node.prop === 'border-right-width') {
                  node.prop = 'border-left-width'
                }
                else if (node.prop === '-webkit-margin-start' || node.prop === '-webkit-margin-end' || node.prop === '-webkit-margin-before' || node.prop === '-webkit-margin-after') {
                  node.remove()
                }
              }
            }
          }
        }
      }
      else if (selector.type === 'combinator') {
        if (selector.value === '>') {
          const nodes = selector.parent?.nodes
          if (nodes) {
            const first = nodes[index + 1]
            if (first && first.type === 'pseudo' && first.value === ':not'
              && (
                (first.first.first.type === 'attribute' && first.first.first.attribute === 'hidden')
                || (first.first.first.type === 'tag' && first.first.first.value === 'template')
              )) {
              const second = nodes[index + 2]
              if (second && second.type === 'combinator' && (second.value === '~' || second.value === '+')) {
                const third = nodes[index + 3]
                if (third && third.type === 'pseudo' && third.value === ':not' && (
                  (third.first.first.type === 'attribute' && third.first.first.attribute === 'hidden')
                  || (third.first.first.type === 'tag' && third.first.first.value === 'template')
                )) {
                  const ast = getCombinatorSelectorAst(options)
                  selector.parent?.nodes.splice(
                    index + 1,
                    3,
                    ...ast,
                  )
                }
              }
            }
          }
        }
      }
      else if (selector.type === 'tag') {
        if (uniAppX) {
          selector.remove()
        }
      }
      else if (selector.type === 'attribute') {
        if (uniAppX) {
          selector.remove()
        }
      }
    })
    selectors.walk((selector) => {
      if (selector.type === 'selector') {
        selector.length === 0 && selector.remove()
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  const parser = psp(transform)

  return (rule: Rule) => {
    currentRule = rule
    try {
      parser.transformSync(rule, normalizeTransformOptions())
    }
    finally {
      currentRule = undefined
    }
  }
}

export function ruleTransformSync(rule: Rule, options: IStyleHandlerOptions) {
  let transformer = ruleTransformCache.get(options)
  if (!transformer) {
    transformer = createRuleTransformer(options)
    ruleTransformCache.set(options, transformer)
  }
  transformer(rule)
}
