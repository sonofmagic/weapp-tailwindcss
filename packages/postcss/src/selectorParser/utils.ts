// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
