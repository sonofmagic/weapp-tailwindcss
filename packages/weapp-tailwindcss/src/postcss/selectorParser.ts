import selectorParser from 'postcss-selector-parser'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { Rule } from 'postcss'
import { composeIsPseudo, internalCssSelectorReplacer } from './shared'
import type { IStyleHandlerOptions } from '@/types'

const createRuleTransform = (rule: Rule, options: IStyleHandlerOptions) => {
  const { escapeMap, mangleContext, cssSelectorReplacement, cssRemoveHoverPseudoClass } = options

  const transform: SyncProcessor = (selectors) => {
    selectors.walk((selector) => {
      // do something with the selector
      // node.selector.replace(/\*/g, 'view')
      if (selector.type === 'universal' && cssSelectorReplacement && cssSelectorReplacement.universal) {
        selector.value = composeIsPseudo(cssSelectorReplacement.universal)
      }

      if (cssRemoveHoverPseudoClass && selector.type === 'selector') {
        const node = selector.nodes.find((x) => x.type === 'pseudo' && x.value === ':hover')
        node && selector.remove()
      }

      if (
        selector.type === 'pseudo' &&
        selector.value === ':root' &&
        cssSelectorReplacement &&
        cssSelectorReplacement.root
      ) {
        selector.value = composeIsPseudo(cssSelectorReplacement.root)
      }

      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
          mangleContext
        })
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  return transform
}

const getRuleTransformer = (rule: Rule, options: IStyleHandlerOptions) => {
  return selectorParser(createRuleTransform(rule, options))
}

export const ruleTransformSync = (rule: Rule, options: IStyleHandlerOptions) => {
  const transformer = getRuleTransformer(rule, options)

  return transformer.transformSync(rule, {
    lossless: false,
    updateSelector: true
  })
}

export function isOnlyBeforeAndAfterPseudoElement(node: Rule) {
  let b = false
  let a = false

  selectorParser((selectors) => {
    selectors.walkPseudos((s) => {
      if (s.parent?.length === 1) {
        if (/^:?:before$/.test(s.value)) {
          b = true
        }
        if (/^:?:after$/.test(s.value)) {
          a = true
        }
      }
    })
  }).astSync(node)

  return b && a
}

export const fallbackRemove = selectorParser((selectors) => {
  selectors.walk((selector) => {
    if (selector.type === 'universal') {
      selector.parent?.remove()
    }
    if (selector.type === 'pseudo' && selector.value === ':is') {
      selector.parent?.remove()
    }
  })
})
