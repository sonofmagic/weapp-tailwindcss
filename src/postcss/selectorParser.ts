import selectorParser from 'postcss-selector-parser'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { Rule } from 'postcss'
import { internalCssSelectorReplacer } from './shared'
import type { IStyleHandlerOptions } from '@/types'

const createTransform = (rule: Rule, options: IStyleHandlerOptions) => {
  const { replaceUniversalSelectorWith, escapeMap, mangleContext, cssSelectorReplacement } = options
  const replaceFlag = replaceUniversalSelectorWith !== false
  const transform: SyncProcessor = (selectors) => {
    selectors.walk((selector) => {
      // do something with the selector
      // node.selector.replace(/\*/g, 'view')
      if (selector.type === 'universal') {
        if (replaceFlag) {
          selector.value = replaceUniversalSelectorWith as string
        } else if (cssSelectorReplacement && cssSelectorReplacement.universal) {
          selector.value = cssSelectorReplacement.universal
        }
      }

      if (selector.type === 'selector') {
        const node = selector.nodes.find((x) => x.type === 'pseudo' && x.value === ':hover')
        node && selector.remove()
      }

      if (selector.type === 'pseudo' && selector.value === ':root' && cssSelectorReplacement && cssSelectorReplacement.root) {
        selector.value = cssSelectorReplacement.root
      }

      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
          mangleContext
        })
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  return transform
}

const getTransformer = (rule: Rule, options: IStyleHandlerOptions) => {
  return selectorParser(createTransform(rule, options))
}

export const transformSync = (rule: Rule, options: IStyleHandlerOptions) => {
  const transformer = getTransformer(rule, options)

  return transformer.transformSync(rule, {
    lossless: false,
    updateSelector: true
  })
}
