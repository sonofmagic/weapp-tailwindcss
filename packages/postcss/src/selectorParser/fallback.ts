// 针对小程序不支持的选择器语法提供兜底清理能力
import type { Rule } from 'postcss'
import type { IStyleHandlerOptions } from '../types'
import type { RuleTransformer } from './rule-transformer'
import type { ParserTransformOptions } from './utils'
import psp from 'postcss-selector-parser'
import { normalizeTransformOptions } from './utils'

const fallbackRemoveCache = new WeakMap<object, {
  parser: ReturnType<typeof psp>
  transform: RuleTransformer
}>()
const fallbackDefaultKey: object = {}

/**
 * 获取用于小程序兼容性处理的解析器，内部会缓存实例并移除不支持的选择器。
 * @returns 带清理规则的解析器实例。
 */
export function getFallbackRemove(_rule?: Rule, options?: IStyleHandlerOptions) {
  const cacheKey = options ?? fallbackDefaultKey
  let entry = fallbackRemoveCache.get(cacheKey)

  if (!entry) {
    const uniAppX = Boolean(options?.uniAppX)
    let currentRule: Rule | undefined

    const parser = psp((selectors) => {
      const activeRule = currentRule
      let maybeImportantId = false
      selectors.walk((selector, idx) => {
        if (idx === 0 && (selector.type === 'id' || selector.type === 'class' || selector.type === 'attribute')) {
          maybeImportantId = true
        }
        if (selector.type === 'universal') {
          selector.parent?.remove()
        }
        else if (selector.type === 'pseudo') {
          if (selector.value === ':is') {
            if (maybeImportantId && selector.nodes[0]?.type === 'selector') {
              selector.replaceWith(selector.nodes[0])
            }
            else {
              selector.parent?.remove()
            }
          }
          else if (selector.value === ':not') {
            for (const x of selector.nodes) {
              if (
                x.nodes.length === 1
                && x.nodes[0].type === 'id'
                && x.nodes[0].value === '#'
              ) {
                x.nodes = [
                  psp.tag({
                    value: '#n',
                  }),
                ]
              }
            }
          }
          else if (selector.value === ':where') {
            for (const n of selector.nodes) {
              for (const node of n.nodes) {
                if (node.type === 'attribute') {
                  node.remove()
                }
              }
            }
          }
        }
        else if (selector.type === 'attribute') {
          if (selector.attribute === 'hidden') {
            activeRule?.remove()
          }
        }
      })
      selectors.walk((selector) => {
        if (selector.type === 'pseudo') {
          if (selector.value === ':where') {
            const res = selector.nodes.every(x => x.nodes.length === 0)
            if (res) {
              selector.remove()
            }
          }
          else if (selector.type === 'pseudo') {
            if (uniAppX) {
              selector.remove()
            }
          }
        }
      })
    })

    const rawTransformSync = parser.transformSync.bind(parser)

    const transform: RuleTransformer = (targetRule: Rule) => {
      currentRule = targetRule
      try {
        rawTransformSync(targetRule, normalizeTransformOptions())
      }
      finally {
        currentRule = undefined
      }
    }

    parser.transformSync = ((input: unknown, opts?: ParserTransformOptions) => {
      if (input && typeof input === 'object' && 'type' in (input as Record<string, unknown>)) {
        const maybeRule = input as Record<string, unknown>
        if (maybeRule.type === 'rule') {
          currentRule = input as Rule
          try {
            return rawTransformSync(input as string | Rule, normalizeTransformOptions(opts))
          }
          finally {
            currentRule = undefined
          }
        }
      }
      return rawTransformSync(input as string | Rule, opts)
    }) as typeof parser.transformSync

    entry = {
      parser,
      transform,
    }
    fallbackRemoveCache.set(cacheKey, entry)
  }

  return entry.parser
}
