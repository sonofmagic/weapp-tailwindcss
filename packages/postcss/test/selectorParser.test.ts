import psp from 'postcss-selector-parser'
import { getFallbackRemove } from '@/selectorParser'

describe('selectorParser', () => {
  it('fallbackRemove case 0', async () => {
    const x = await getFallbackRemove().process('#app-provider :is(.space-x-4>view+view)')
    expect(x).toMatchSnapshot()
  })

  it('fallbackRemove case 1', async () => {
    const x = await getFallbackRemove().process('#app :is(.space-x-0>view+view)')
    expect(x).toMatchSnapshot()
  })

  it('fallbackRemove case 2', async () => {
    const x = await getFallbackRemove().process('button:not(n),input:where(,,):not(n)')
    expect(x).toMatchSnapshot()
  })

  it('fallbackRemove case 3', async () => {
    const x = await getFallbackRemove().process('[hidden]:where(:not([hidden="until-found"])):not(n):not(n):not(n)')
    expect(x).toMatchSnapshot()
  })

  it('fallbackRemove case 4', async () => {
    const x = await getFallbackRemove().process('button:not(n),input:where():not(n)')
    expect(x).toMatchSnapshot()
  })

  it('ruleTransformSync case 0', async () => {
    const target = psp((selectors) => {
      selectors.walkTags((tag) => {
        tag.remove()
      })
    })
    const normalized = await target.process('.dark view.darkcbg-black,.dark text.darkcbg-black', { updateSelector: true })
    expect(normalized).toMatchSnapshot()
  })
})
