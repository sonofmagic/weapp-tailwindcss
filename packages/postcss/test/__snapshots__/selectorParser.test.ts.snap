// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selectorParser > fallbackRemove case 0 1`] = `"#app-provider .space-x-4>view+view"`;

exports[`selectorParser > fallbackRemove case 1 1`] = `"#app .space-x-0>view+view"`;

exports[`selectorParser > fallbackRemove case 2 1`] = `"button:not(n),input:not(n)"`;

exports[`selectorParser > fallbackRemove case 3 1`] = `"[hidden]:where(:not([hidden="until-found"])):not(n):not(n):not(n)"`;

exports[`selectorParser > fallbackRemove case 4 1`] = `"button:not(n),input:not(n)"`;

exports[`selectorParser > ruleTransformSync case 0 1`] = `".dark .darkcbg-black,.dark .darkcbg-black"`;
